### ISSUE TITLE
xgboost 진로성숙도 파라미터 튜닝

### ISSUE THREAD

**Issue by refree** [refree/parameter-tunning#5]
#xgboost는 caret 에서 파라미터 튜닝할때 분류의 경우 목표변수 자료유형 Factor / 나머지 numeric
#진짜 xgboost에서는 목표변수 integer / 나머지 numeric
#참고 블로그(blog) https://datascience-enthusiast.com/R/ML_python_R_part2.html

#진로성숙도 Xgboost 파라미터 뉴닝

setwd("C:/Users/Park Jung/Desktop/SELS/data/query")
library(caret); library(dplyr); library(randomForest);
library(xgboost)
cb7 <- read.csv("7grade_join_1.csv" )
cb8 <- read.csv("8grade_join_1.csv" );cb9 <- read.csv("9grade_join_1.csv" )

summary(cb7$career_maturity)
summary(cb8$career_maturity);summary(cb9$career_maturity)
#- numeric --------#
for(i in 1:length(cb7)) { cb7[,i] <- as.numeric(cb7[,i]) }
for(i in 1:length(cb8)) { cb8[,i] <- as.numeric(cb8[,i]) }; for(i in 1:length(cb9)) { cb9[,i] <- as.numeric(cb9[,i]) }
sum(is.na(cb7)); cb7 <- as.data.frame(cb7);
sum(is.na(cb8)); cb8 <- as.data.frame(cb8);sum(is.na(cb9)); cb9 <- as.data.frame(cb9);

cb7 <- cb7 %>%  mutate(target = ifelse(career_maturity >= 0.7482, 1, 0))
cb8 <- cb8 %>%  mutate(target = ifelse(career_maturity >= 0.7222, 1, 0)); cb9 <- cb9 %>%  mutate(target = ifelse(career_maturity >= 0.7342, 1, 0))
str(cb7)
summary(cb8$career_maturity);summary(cb9$career_maturity)

cb7$target <- as.factor(cb7$target)
cb8$target <- as.factor(cb8$target); cb9$target <- as.factor(cb9$target)
cb7$career_maturity <- NULL;
cb8$career_maturity <- NULL;cb9$career_maturity <- NULL;
summary(cb8$target);summary(cb9$target)

library(Matrix)
#xgboost 파라미터 세팅 코드
X_train = xgb.DMatrix(as.matrix(cb7 %>% select(-target)))
y_train = cb7$target

X_train8 = xgb.DMatrix(as.matrix(cb8 %>% select(-target)))
y_train8 = cb8$target

X_train9 = xgb.DMatrix(as.matrix(cb9 %>% select(-target)))
y_train9 = cb9$target

xgb_trcontrol = trainControl(method = "cv", number = 5,
                             allowParallel = TRUE, verboseIter = FALSE,
                             returnData = FALSE )

xgbGrid <- expand.grid(nrounds = c(100,200), # this is n_estimators in the python code above
                       max_depth = c(10, 15, 20, 25),
                       colsample_bytree = seq(0.5, 0.9, length.out = 5),
                       ## The values below are default values in the sklearn-api.
                       eta = 0.1, #학습률은 부스팅 단계마다 가중치를 줘서 부스팅 과정에 오버피팅이 일어나지 않도록 한
                       gamma=0, # 값 5 정보 획득량에서 로 표현한 값이며, 감마가 커질수록 트리의 깊이가 줄어들어 보수적인 모형이 된
                       min_child_weight = 1,
                       subsample = 1 )

set.seed(0)
#modelLookup(model='xgbTree')
system.time(xgb_model <- train(X_train, y_train, trControl = xgb_trcontrol,
                               metric = "Accuracy", tuneGrid = xgbGrid, method = "xgbTree"))
print(xgb_model); plot(xgb_model)

system.time(xgb_model8 <- train(X_train8, y_train8, trControl = xgb_trcontrol,
                               metric = "Accuracy", tuneGrid = xgbGrid, method = "xgbTree"))
print(xgb_model8); plot(xgb_model8)

system.time(xgb_model9 <- train(X_train9, y_train9, trControl = xgb_trcontrol,
                               metric = "Accuracy", tuneGrid = xgbGrid, method = "xgbTree"))
print(xgb_model9); plot(xgb_model9)
